### ISSUE TITLE
Split abstract TD and protocol binding

### ISSUE THREAD

**Issue by mlagally** [w3c/wot-thing-description#91]
The combination of the abstract interaction model with the protocol bindings for different protocols has several disadvantages:

1. Difficult maintenance
When a new protocol is added or a bug is fixed for a single protocol, the change affects all users of the TD, even if they don't use that protocol at all
2. Difficult extensibility for other protocols
If another protocol for specific use cases must be supported, the TD needs to be changed
It is unclear how additional vocabulary is handled in this case
3. Parsing overhead:
A resource constrained device needs, which has only support for a single protocol, has to parse a large TD to get the protocol bindings, that are relevant for it

A clean separation of the interaction model and the different protocol bindings into separate documents would address these issues

**Comment by mkovatsc** [comment]
We already have a logical split. All binding information is under `form`. Maybe we should rename it to`binding` to make it clear -- and release Mozilla from thinking about the difference between links and forms :P

The protocol vocabulary is also defined in a separate file: a Binding Template. Putting the instance information it into a separate document than the TD puts us back into integration hell: incomplete metadata, multiple roundtrips to collect information. When you want to have it concise, simply align your implementation with the defaults which can boil the binding part down to an href.

**Comment by sebastiankb** [comment]
We will stick with 'form' based on decission on the last WoT meeting on Wednesday.

All protocol related metadata should be placed in the form container. As Matthias mentioned, there is already this split of the abstraction interaction model  and the concrete protocol used.


Can we close this issue?

**Comment by mkovatsc** [comment]
Maybe another clarification: We currently did some work on the data schema in the Binding Templates Task Force, because there we had a number of real-world examples how Interaction resources would look like.

The data schema is part of the TD and describes the schema at Interaction level (Property, Action, Event). It will be fully specified in the TD spec. At the moment we are only referencing the extensive examples in the Binding Templates document. It will just show how the schema would describe OCF, IPSO Objects, etc.

For the protocol-specific part, TD has the `form` extension point. I expect a small set of defaults for HTTP, CoAP, and something like Server-Sent Events or pure Event data push over a raw WebSocket. Everything that goes beyond that is handled outside TD with an initial set of examples in the Binding Templates document.

Not to actual TDs themselves: a TD shall be self-contained meaning it should be able to carry all the metadata needed to interact with Things (modulo the defaults mentioned such as GET for read, POST for invoke...). Externalizing this is always possible with Linked Data. Yet before we do that, there should be a common understanding of what these agreed-on out-of-band pieces of metadata should be. To get started, I would expect that these TDs with custom metadata will not be shared and only serve for the internal integration, e.g., consuming BACnet, Modbus, and Echonet Things and exposing them with the default protocols, whereas those ExposedThing TDs are shared. The advantage of this is that semantic metadata can be attached directly at the "weird devices" such as Modbus slaves and easily forwarded at the gateway. We do not have the usual information loss due to translating gateways*.

*Everyone who has worked on IoT for a while knows that these translating gateways are a big pain and do not scale if they are based on individual translations without a common concept.

@mlagally Does this cover your concern?

**Comment by mlagally** [comment]
Hi Matthias, Hi Sebastian,

unfortunately I'm still concerned of mixing the logical thing description with the protocol binding for specific protocols in the same file.

From a deployment perspective this complicates things. If you think of a thing description that is used by two different device manufacturers who agree on the same (semantic) model but implement different (potentially proprietary) protocols, somebody has to author a combined TD.
Who will do that? Manufacturer A? Manufacturer B? Both?

Most likely both manufacturers will author their own TD, and most likely they won't integrate it, since the protocol binding of the manufacturer B is useless for the device of manufacturer A.

This will cause fragmentation of the TDs and hinder interoperability, which would only be possible if both of them add semantic annotations to the model. However if one of them doesn't, there is an unnecessary duplication of a common thing model and no interoperability.

Therefore it makes sense to keep the protocol binding separate. From a maintenance perspective, each binding to a specific protocol should be in a self-contained document.

**Comment by mkovatsc** [comment]
But complete TDs are always instance-specific, as they contain at least the URIs with the network addresses for  each specific Thing as well as identities. In that form, they won't be shared across devices.

Thus, there is the idea of Thing templates, basically a TD without the instance-specific metadata, i.e., now form/link elements, and hence no bindings. This what would be equivalent to Oracle's Device Model.

There are more subsets of a TD (are smaller Shapes over the triples that describe a Thing..). Kajimoto-san had something on this that also supports the design process of industry standards, product lines, models, etc. We will discuss these lifecycle aspects again in Prague.

**Comment by benfrancis** [comment]
@mlagally wrote:
> From a maintenance perspective, each binding to a specific protocol should be in a self-contained document.

I agree, but in the sense that the protocol binding could be defined in a formal protocol binding specification outside the Thing Description rather than trying to somehow describe each protocol declaratively using metadata.

> If you think of a thing description that is used by two different device manufacturers who agree on the same (semantic) model but implement different (potentially proprietary) protocols, somebody has to author a combined TD.

I'm curious what you think such a Thing Description would actually achieve? Does sharing the same semantic model really help interoperability much if the two things use incompatible proprietary protocols?

What if all web things used the same (web) protocol at the Web of Things layer? A gateway application would be used to translate those two proprietary protocols into the standard web protocol and a WoT client could then talk to both web things using a single standard protocol.

**Comment by mlagally** [comment]
@mkovatsc : I agree - we should have a template mechanism, which describes the properties, actions and events. We could use the proposed (perhaps simplified) TD format as a starting point, which is the blueprint for things that adopt the model.
This blueprint should ideally not deal with protocols - perhaps a default binding to HTTP/REST could be defined.

@benfrancis: if not in a declarative description, how would you define the protocol binding?
There is one scenario, where it could be useful to just have a common semantic model: If you want to create an application, based on the scripting API, you could deploy it on both devices, where it shows the same behaviour. This would increase the target market for application developers.

**Comment by benfrancis** [comment]
> @benfrancis: if not in a declarative description, how would you define the protocol binding?

In a protocol binding specification, like [this one](https://iot.mozilla.org/wot/#web-thing-rest-api), or [this one](https://iot.mozilla.org/wot/#web-thing-websocket-api), but more formally defined as a standards track document.

> There is one scenario, where it could be useful to just have a common semantic model: If you want to create an application, based on the scripting API, you could deploy it on both devices, where it shows the same behaviour. This would increase the target market for application developers.

I'd argue that standardising a scripting API is unnecessary if the Web of Things converges on a small number of web protocols. Browsers and other web clients already have scripting APIs to talk HTTP and WebSockets. No browser vendor has expressed an interest in implementing the WoT Scripting API and I don't see the point in trying to define a language-agnostic scripting API. Every library and framework will want to do it differently and should be allowed to compete. In my view it's better to standardise at the protocol level.

**Comment by mkovatsc** [comment]
@mlagally could you provide your list of "types of required TDs" here that you mentioned in a recent call? I think your split of concerns is quite related to them.

As explained earlier, TD templates do not contain binding information (no forms or security metadata). The TD templates are used for abstract Things (~classes), for instance, in the Scripting API to instantiate Things of that class. The runtime, which knows about available bindings and the security mechanisms would fill in the corresponding information; the templates become TDs describing the Thing instances.

**Comment by mlagally** [comment]
@mkovatsc I'm not sure I remember my exact comment about "types of required TSs". Basically if we have a "OO class-like" blueprint mechanism for a thing-type TD and a mechanism for describing an instance 
-TD we are good.
